feat: add storage control anywhere cache samples

thiyaguk09 opened the pull request:
## Description

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [ ] I have followed guidelines from [CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md) and [Samples Style Guide](https://googlecloudplatform.github.io/samples-style-guide/)
- [ ] **Tests** pass:   `npm test` (see [Testing](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#run-the-tests-for-a-single-sample))
- [ ] **Lint** pass:   `npm run lint` (see [Style](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#style))
- [ ] **Required CI tests** pass (see [CI testing](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#ci-testing))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This pull request is from a branch created directly off of `GoogleCloudPlatform/nodejs-docs-samples`. Not a fork.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new sample directory, and I created [GitHub Actions workflow](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#adding-new-samples) for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [ ] Please **merge** this PR for me once it is approved

> **Note**: Any check with `(dev)`, `(experimental)`, or `(legacy)` can be ignored and should **not block** your PR from merging (see [CI testing](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#ci-testing)).

thiyaguk09 committed:
feat(storagecontrol): Add samples for Storage Anywhere Cache

thiyaguk09 committed:
add testcase

thiyaguk09 committed:
lint fix

thiyaguk09 committed:
fix: test case

thiyaguk09 committed:
fix: remove cacheName

thiyaguk09 committed:
skip test case

thiyaguk09 committed:
test case remove projectId

iennae commented on storage-control/createAnywhereCache.js:
```diff
@@ -0,0 +1,102 @@
+
+function main(bucketName, zoneName) {
+  // [START storage_control_create_anywhere_cache]
+  /**
+   * TODO(developer): Uncomment these variables before running the sample.
+   */
```
suggestion: 

/**
 * Creates an Anywhere Cache instance for a Cloud Storage bucket.
 *
 * Anywhere Cache is a feature that provides an SSD-backed zonal read cache.
 * This can significantly improve read performance for frequently accessed data
 * by caching it in the same zone as your compute resources.
 *
 * @param {string} bucketName The name of the bucket to create the cache for.
 *   Example: 'your-gcp-bucket-name'
 * @param {string} zone The zone where the cache will be created.
 *   Example: 'us-central1-a'
 */

thiyaguk09 replied:
Is there a standard currently being followed that we should adhere to for future reference?

iennae commented on storage-control/createAnywhereCache.js:
```diff
@@ -0,0 +1,102 @@
+    const request = {
+      parent: bucketPath,
+      anywhereCache: {
+        zone: zoneName,
```
should the default sample show a TTL?

iennae commented on storage-control/disableAnywhereCache.js:
```diff
@@ -0,0 +1,60 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+function main(bucketName, cacheName) {
+  // [START storage_control_disable_anywhere_cache]
+  /**
+   * TODO(developer): Uncomment these variables before running the sample.
+   */
```
suggestion: 

/**
 * Disables an Anywhere Cache instance.
 *
 * Disabling a cache is the first step to permanently removing it. Once disabled,
 * the cache stops ingesting new data. After a grace period, the cache and its
 * contents are deleted. This is useful for decommissioning caches that are no
 * longer needed.
 *
 * @param {string} bucketName The name of the bucket where the cache resides.
 *   Example: 'your-gcp-bucket-name'
 * @param {string} cacheName The unique identifier of the cache instance to disable.
 *   Example: 'cacheName'
 */

thiyaguk09 replied:
Thank you for your thoughtful suggestion and for providing such a well-crafted JSDoc comment. We aim to maintain a consistent documentation style across all our Cloud Storage samples to offer a unified experience for our users.

Adopting a new format, even if it is an improvement, could impact the overall cohesiveness and clarity of the sample set.

thiyaguk09 replied:
Thank you again for the suggestion.

I've recently received confirmation regarding the new structural changes being applied across all SDKs and samples. Since we are moving toward this new documentation style globally, **your suggested format is now appropriate and should be adopted here** to align with the future direction.

iennae commented on storage-control/disableAnywhereCache.js:
```diff
@@ -0,0 +1,60 @@
+    };
+
+    // Run request
+    const [response] = await controlClient.disableAnywhereCache(request);
```
This will initiate the disablement right? and there is some grace period where it can be recovered?

thiyaguk09 replied:
Yes, you're correct. The action will **immediately change the state** of the cache to disabled. There is no grace period for recovery; once you initiate the disablement, it is a final, non-reversible process.

The actual removal of the cache's data and resources happens in the background. Unlike `create` or `update` operations, which can be tracked with a long-running operation ID, disabling a cache is a direct state change without a corresponding operation.

iennae commented on storage-control/getAnywhereCache.js:
```diff
@@ -0,0 +1,60 @@
+
+function main(bucketName, cacheName) {
+  // [START storage_control_get_anywhere_cache]
+  /**
```
nit: suggestion of 

/**
 * Retrieves details of a specific Anywhere Cache instance.
 *
 * This function is useful for checking the current state, configuration (like TTL),
 * and other metadata of an existing cache.
 *
 * @param {string} bucketName The name of the bucket where the cache resides.
 *   Example: 'your-gcp-bucket-name'
 * @param {string} cacheName The unique identifier of the cache instance.
 *   Example: 'my-anywhere-cache-id'
 */

thiyaguk09 replied:
Thank you for your thoughtful suggestion and for providing such a well-crafted JSDoc comment. We aim to maintain a consistent documentation style across all our Cloud Storage samples to offer a unified experience for our users.

Adopting a new format, even if it is an improvement, could impact the overall cohesiveness and clarity of the sample set.

thiyaguk09 replied:
Thank you again for the suggestion.

I've recently received confirmation regarding the new structural changes being applied across all SDKs and samples. Since we are moving toward this new documentation style globally, **your suggested format is now appropriate and should be adopted here** to align with the future direction.

iennae commented on storage-control/getAnywhereCache.js:
```diff
@@ -0,0 +1,60 @@
+
+    // Run request
+    const [response] = await controlClient.getAnywhereCache(request);
+    console.log(`Got anywhere cache: ${response.name}.`);
```
nit: beyond just the name, might be helpful to show some specific data that is responded. 

    console.log(`Anywhere Cache details for '${cacheName}':`);
    console.log(`  ID: ${response.id}`);
    console.log(`  Zone: ${response.zone}`);
    console.log(`  State: ${response.state}`);
    console.log(`  TTL: ${response.ttl.seconds}s`);
    console.log(`  Admission Policy: ${response.admissionPolicy}`);
    console.log(`  Create Time: ${new Date(response.createTime.seconds * 1000).toISOString()}`);

thiyaguk09 replied:
For get responses, it now displays additional data like this.

```
console.log(`Anywhere Cache details for '${cacheName}':`);
console.log(`  Name: ${response.name}`);
console.log(`  Zone: ${response.zone}`);
console.log(`  State: ${response.state}`);
console.log(`  TTL: ${response.ttl.seconds}s`);
console.log(`  Admission Policy: ${response.admissionPolicy}`);
console.log(`  Create Time: ${new Date(response.createTime.seconds * 1000).toISOString()}`);
```

iennae commented on storage-control/updateAnywhereCache.js:
```diff
@@ -0,0 +1,71 @@
+
+function main(bucketName, cacheName, admissionPolicy) {
+  // [START storage_control_update_anywhere_cache]
+  /**
```
nit: 

/**
 * Updates the Admission Policy of an Anywhere Cache instance.
 *
 *
 * @param {string} bucketName The name of the bucket where the cache resides.
 *   Example: 'your-gcp-bucket-name'
 * @param {string} cacheName The unique identifier of the cache instance to update.
 *   Example: 'my-anywhere-cache-id'
 *  @param {string} admissionPolicy Determines when data is ingested into the cache
 */

thiyaguk09 replied:
Thank you for your thoughtful suggestion and for providing such a well-crafted JSDoc comment. We aim to maintain a consistent documentation style across all our Cloud Storage samples to offer a unified experience for our users.

Adopting a new format, even if it is an improvement, could impact the overall cohesiveness and clarity of the sample set.

iennae commented on storage-control/updateAnywhereCache.js:
```diff
@@ -0,0 +1,71 @@
+  // const cacheName = 'cacheName';
+
+  // The admission policy of the cache to be updated
+  // const admissionPolicy = 'admit-on-first-miss';
```
by default this looks like the default setting. might suggest something else so it changes? this is based on reading https://cloud.google.com/storage/docs/anywhere-cache

thiyaguk09 replied:
For `admissionPolicy`, we can set it to either `admit-on-first-miss` or `admit-on-second-miss`. The default is `admit-on-first-miss`, but here I have changed it to `admit-on-second-miss`.

iennae reviewed (commented):
This mostly looks ok. I have a few questions. I did run this through our generation to see what kind of output I received and I've shared those headers for visibility. I think there are a few areas of improvements to make sure there is a try..catch to ensure folks get actionable feedback so they don't assume the code is broken depending on the order in which they run commands. 

Thanks!

thiyaguk09 commented on storage-control/disableAnywhereCache.js:
```diff
@@ -0,0 +1,60 @@
+
+    // Run request
+    const [response] = await controlClient.disableAnywhereCache(request);
+    console.log(`Disabled anywhere cache: ${response.name}.`);
```
However, our code samples are designed to demonstrate the simplest, **"happy path"** usage of a feature. We intentionally omit error handling to keep the code concise and focused on the core API call. This approach ensures all Cloud Storage samples share a consistent, minimal style, making them easier to understand at a glance.

thiyaguk09 replied:
I have added a try-catch block and listed the possible errors in the catch block.

thiyaguk09 committed:
addressing review comments

thiyaguk09 commented:
> This mostly looks ok. I have a few questions. I did run this through our generation to see what kind of output I received and I've shared those headers for visibility. I think there are a few areas of improvements to make sure there is a try..catch to ensure folks get actionable feedback so they don't assume the code is broken depending on the order in which they run commands.
> 
> Thanks!

Thank you for the suggestion. Our samples are designed to demonstrate the simplest "happy path" for a feature, focusing on the core API call. We intentionally omit try...catch blocks to keep the code concise and readable, which helps maintain a consistent style across all of our language samples. Furthermore, unhandled rejections are already handled at the application level to prevent the code from breaking, so the primary goal of our samples remains focused on positive-case demonstration.

thiyaguk09 committed:
index on anywhere-cache-samples: e83ba77b addressing review comments

thiyaguk09 committed:
feat(storage-control): Improve Anywhere Cache API samples

Improve Anywhere Cache API samples (Documentation & Error Handling)

* Wraps all asynchronous Anywhere Cache API samples (Create, Get, List,
Disable, Pause, Resume) in `try...catch` blocks for production
readiness.
* Adds specific gRPC error code checks (NOT_FOUND, FAILED_PRECONDITION)
to provide better diagnostic feedback to users.
* Clarifies documentation regarding the optional nature and default
values of `ttl` and `admissionPolicy` for cache creation.

thiyaguk09 commented:
@iennae That's great feedback, thank you for running the samples and sharing the output headers—having that visibility is really helpful. The issues have been fixed and pushed here; could you please take a look?

thiyaguk09 committed:
test(storagecontrol): Enhance Anywhere Cache test assertions and coverage

Updates tests for all Anywhere Cache management methods (Create, Get,
List, Resume, Disable, Pause) to align with enhanced sample script
output.

**(Tests were skipped due to reliance on Long-Running Operations (LROs)
in the sample code.)**

This includes:
* **Enhanced Assertions:** Asserting against all newly added detail
fields (e.g., Name, State, TTL) to verify full API response parsing in
Get and List samples.
* **Negative Scenario Coverage:** Adding explicit **negative scenario
tests** for state-dependent operations (Disable, Pause, Resume) to
assert graceful failure on expected errors like `FAILED_PRECONDITION`.

thiyaguk09 committed:
Fix: Remove duplicate call to disableAnywhereCache

The function was being called twice due to an oversight during a recent try/catch restructure.

thiyaguk09 commented on storage-control/disableAnywhereCache.js:
```diff
@@ -0,0 +1,104 @@
+      throw error;
+    }
+    // Run request
+    const [response] = await controlClient.disableAnywhereCache(request);
```
Thanks for spotting this! You're absolutely right; calling `disableAnywhereCache` twice is unnecessary.

It seems I introduced that redundancy in a recent commit during a **restructuring with a `try...catch` block**, and I missed deleting the extra lines.

I've **fixed this now** and removed the duplicate call.

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

glasnt committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples

chandra-siri committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

thiyaguk09 committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples

iennae committed:
Merge branch 'main' into anywhere-cache-samples
